models/cross-spork_client.go

```diff
@@ -0,0 +1,151 @@
+	logger zerolog.Logger
+	// this map holds the last heights and clients for each spork
+	sporkHosts map[uint64]*grpc.BaseClient
+	*grpc.BaseClient
+}
+
+// NewCrossSporkClient creates a new instance of the client, it accepts the
+// host to the current spork AN API.
+func NewCrossSporkClient(currentSporkHost string, logger zerolog.Logger) (*CrossSporkClient, error) {
+	// add current spork AN host as the default client
+	client, err := grpc.NewBaseClient(currentSporkHost, dialOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	return &CrossSporkClient{
+		logger,
+		make(map[uint64]*grpc.BaseClient),
+		client,
+	}, nil
+}
+
+// AddSpork will add a new spork host defined by the last height boundary in that spork.
+func (c *CrossSporkClient) AddSpork(lastHeight uint64, host string) error {
+	if _, ok := c.sporkHosts[lastHeight]; ok {
+		return fmt.Errorf("provided last height already exists")
+	}
+
+	client, err := grpc.NewBaseClient(host, dialOpts)
+	if err != nil {
+		return err
+	}
+
+	c.sporkHosts[lastHeight] = client
+
+	c.logger.Info().
+		Uint64("spork-boundary", lastHeight).
+		Str("host", host).
+		Msg("added spork specific client")
+
+	return nil
+}
+
+// IsPastSpork will check if the provided height is contained in the previous sporks.
+func (c *CrossSporkClient) IsPastSpork(height uint64) bool {
+	return len(c.getSporkBoundariesDesc()) > 0 && height <= c.getSporkBoundariesDesc()[0]
+}
+
+// getClientForHeight returns the client for the given height. It starts by using the current spork client,
+// then iteratively checks the upper height boundaries in descending order and returns the last client
+// that still contains the given height within its upper height limit. If no client is found, it returns
+// the current spork client.
+// Please note that even if a client for provided height is found we don't guarantee the data being available
+// because it still might not have access to the height provided, because there might be other sporks with
+// lower height boundaries that we didn't configure for.
+// This would result in the error when using the client to access such data.
+func (c *CrossSporkClient) getClientForHeight(height uint64) *grpc.BaseClient {
+
+	// start by using the current spork client, then iterate all the upper height boundaries
+	// and find the last client that still contains the height in its upper height limit
+	client := c.BaseClient
+	for _, upperBound := range c.getSporkBoundariesDesc() {
+		if upperBound >= height {
+			client = c.sporkHosts[upperBound]
+
+			c.logger.Debug().
+				Uint64("spork-boundary", upperBound).
+				Msg("using previous spork client")
+		}
+	}
+
+	return client
+}
+
+// getSporkBoundaries will return descending order of spork height boundaries
+func (c *CrossSporkClient) getSporkBoundariesDesc() []uint64 {
+	heights := maps.Keys(c.sporkHosts)
+	slices.Sort(heights)    // order heights in ascending order
+	slices.Reverse(heights) // make it descending
+	return heights
```

Comment:
since `sporkHosts` is never changed during runtime, this could be built once and cached.

Reply:
I know, but I don't think it's worth caching something that will likely be a slice with 5 items. I don't like adding something for sake of performance if it's not justified. If this was a long list maybe, but caching something with less than 10 items I feel is not worth it even if it's 2 lines more.

Reply:
it's not about lines of code. this is just something that gets called on every query. if you don't think there will be many queries, or want to wait to see if this is needed, that's fine.

Reply:
I am also in favor of making the `AddSpork` method non-exportable, and have the `NewCrossSporkClient` method accept just the current spork and past spork hosts, create the clients and do the necessary wiring. That could also include builting once the `heights` map. Relevant comment: https://github.com/onflow/flow-evm-gateway/pull/230#discussion_r1598100266